### ISSUE TITLE
API proposal: New Timer based on IAsyncEnumerable

### ISSUE THREAD

**Issue by davidfowl** [dotnet/corefx#42645]
To solve of the common issues with timers:
1. The fact that it always captures the execution context is problematic for certain long lived operations (https://github.com/dotnet/corefx/issues/32866)
2. The fact that is has strange rooting behavior based on the constructor used
3. The fact timer callbacks can overlap (https://github.com/dotnet/corefx/issues/39313)
4. The fact that timer callbacks aren't asynchronous which leads to people writing sync over async code

I propose we create a modern timer API based on `IAsyncEnumerable<T>` that basically solves all of these problems 😄 .

- This API only makes sense for timers that fire repeatedly, timers that fire once could be Task based (we already have `Task.Delay` for this)
- If user code takes longer to run than the interval, we drop the tick events that happened until the user can catch up (think bounded channel with drop and a capacity of 1).
- The timer will be disposed when the `IAsyncEnumerator<T>` is disposed (not sure if this is good enough)

```C#
public class Timer
{
    public static IAsyncEnumerable<TimerSpan> Create(TimeSpan dueTime, TimeSpan period);
}
```

Usage:

```C#
class Program
{
    static async Task Main(string[] args)
    {
        var second = TimeSpan.FromSeconds(1);
        var ticker = Timer.CreateTimer(second, second);

        await foreach(var t in ticker)
        {
            Console.WriteLine($"Tick {t}")
        }
    }
}
```

cc @stephentoub

**Comment by ehouarn-perret** [comment]
I like the idea, but when it comes to stop the timer I found it a little bit weird (I mean I guess in your example, that would involve a break to get out of the `foreach` loop).

**Comment by davidfowl** [comment]
> I like the idea, but when it comes to stop the timer I found it a little bit weird (I mean I guess in your example, that would involve a break to get out of the foreach loop).

Right, it's pull so you stop the timer by breaking out of the loop, or using a cancellation token with the IAsyncEnumerable to stop it from running (that also works if you're not directly in the loop).

**Comment by NtFreX** [comment]
```
public static class Timer
{
    public static async Task Subscribe(this IAsyncEnumerable<TimerSpan> ticker, Action<TimerSpan> action)
    {
        await foreach(var t in ticker)
        {
            action?.Invoke(t);
        }
    }
}

var second = TimeSpan.FromSeconds(1);
using(var ticker = Timer.CreateTimer(second, second))
{
    await ticker.Subscribe(t => Console.WriteLine($"Tick {t}"));
}
```

As long as we get some good helper methods simular to the one above I like fanciness.

**Comment by davidfowl** [comment]
The entire point of this is to avoid callbacks, but if you want to write that helper, have at it 😄

**Comment by NicolasDorier** [comment]
That is sexy. Though it is difficult to know from the look of it, what happen if a tick happen while we are not `awaiting`. Do they queue up? get ignored?

**Comment by ankitbko** [comment]
> ```
> public static class Timer
> {
>     public static async Task Subscribe(this IAsyncEnumerable<TimerSpan> ticker, Action<TimerSpan> action)
>     {
>         await foreach(var t in ticker)
>         {
>             action?.Invoke(t);
>         }
>     }
> }
> 
> var second = TimeSpan.FromSeconds(1);
> using(var ticker = Timer.CreateTimer(second, second))
> {
>     await ticker.Subscribe(t => Console.WriteLine($"Tick {t}"));
> }
> ```
> 
> As long as we get some good helper methods simular to the one above I like fanciness.

Seems very similar to [rx timer](https://www.learnrxjs.io/operators/creation/timer.html)

**Comment by willdean** [comment]
With apologies for the bike-shedding, but can I suggest it's called something other than `Timer`?

We already have these three famous ones:
`System.Threading.Timer`
`System.Timers.Timer`
`System.Windows.Forms.Timer`
Plus another few I didn't know about until I looked at Reference Source just now.

People don't always qualify which one they're talking about (#32866 is a good example), and adding another new one just makes things hard to learn for newcomers and hard to talk about.

If it was called Timer2020 then as long as there wasn't more than one new timer classed added each year, that would be a scaleable naming scheme...  (Not a serious suggestion, in case my audience is unappreciative.)

**Comment by svick** [comment]
```c#
public class Timer
```

I take it this is [`System.Threading.Timer`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.timer.-ctor?view=netcore-3.0), and not [`System.Timers.Timer`](https://docs.microsoft.com/en-us/dotnet/api/system.timers.timer?view=netcore-3.0), or some new namespace?

```c#
public static IAsyncEnumerable<TimerSpan> Create(TimeSpan dueTime, TimeSpan period);
```

Assuming this method is indeed added to `System.Threading.Timer`, I feel like calling it `Create` is not sufficient to differentiate it from the `Timer` constructor. Though I'm not sure what name would be better here.

---

In general, I agree that having `await`-based timer is useful, but I'm not sure `IAsyncEnumerable<T>` is the right abstraction for that.

For example, consider that I want to output items from a collection to the console, but only at a rate of one per second. With a different design, that could be easily achieved by something like:

```c#
using var timer = Timer.Create(TimeSpan.Zero, TimeSpan.FromSeconds(1));

foreach (var item in items)
{
    await timer.WaitForNext();
    
    Console.WriteLine(item);
}
```

But I think this would be much harder to write using an `IAsyncEnumerable<T>`-based timer, at least without Ix.NET.

**Comment by vshapenko** [comment]
Maybe this should be an extension to System.Threading.Channels?

**Comment by davidfowl** [comment]
@svick you can use IAsyncEnumerable for that as well. What you wrote is literally GetEnumertor + MoveNextAsync

**Comment by alefranz** [comment]
> * If user code takes longer to run than the interval, we drop the tick events that happened until the user can catch up (think bounded channel with drop and a capacity of 1).

That's definitely the right approach, but I think it would be nice to customize what happens when a tick is missed: return immediately on await vs wait for the next tick (so to get back in sync)

**Comment by stephentoub** [comment]
Just to make sure I understand the proposal, you're basically proposing this, right? (I'm not sure what the TimeSpan is meant to be an offset from, so I've just used DateTime here.)
```C#
public static IAsyncEnumerable<DateTime> Create(TimeSpan dueTime, TimeSpan period)
{
    var c = Channel.CreateBounded<DateTime>(1);
    using (var t = new Timer(s => ((Channel<DateTime>)s).Writer.TryWrite(DateTime.UtcNow), c, dueTime, period))
    {
        await foreach (DateTime dt in c.Reader.ReadAllAsync())
            yield return dt;
    }
}
```
?

**Comment by ankitbko** [comment]
If we are open for new name, I would propose `Interval`.

But adding something similar to Timer under new name smells bad. Creates unnecessary confusion.

**Comment by omariom** [comment]
Now we need this
```C#
ChannelReader<Message> channel = GetReader();
var ticker = Timer.CreateTimer(second, second);

(channel, ticker) switch 
{
    TimeSpan tick => 
    {
    },
    Message mesg => 
    {
    }
}
```

I can't find @stephentoub's proposal on the subject :neutral_face:

**Comment by davidfowl** [comment]
@stephentoub Yep. But with `BoundedChannelFullMode.DropOldest`, though that really only matters if we find something useful to put in the `<T>`. Originally I was thinking time since last tick so if the code execution lasts longer than the period, you'd get a timespan that represnted how long between MoveNext calls but it was late, I dunno if that's useful 😄

**Comment by stephentoub** [comment]
> Now we need this

That's problematic with `IAsyncEnumerable<T>`, assuming you want to perform this operation more than once, because a) you need to get an enumerator in order to read from it and that enumerator may "start over" depending on the source, but more importantly b) once you call MoveNext() you've pushed the enumerator forward and you can't take it back; that effectively ruins it for a subsequent consumer.

With channels it's feasible, because rather than MoveNext/Current, the operations are WaitForReadAsync/TryRead, which means unlike MoveNext, asking to be notified of an available item doesn't consume said item.

**Comment by davidfowl** [comment]
> That's definitely the right approach, but I think it would be nice to customize what happens when a tick is missed: return immediately on await vs wait for the next tick (so to get back in sync)

Why?

**Comment by willdean** [comment]
> Why?

I can't speak for the OP, nor whether this sort of an option would acutally be a good thing, but sometimes I want a one-second timer to mean "run this code 3600 times an hour", and sometimes I want it to mean "wake this code up roughly once a second".  

Clearly all such things are no better than 'best effort' on this sort of platform, but to me there is a discernible difference in intention.

**Comment by scalablecory** [comment]
Reactive's `Observable.Interval` is nice, but it seems weird usability-wise for the return to be an `IAsyncEnumerable`.

**Comment by davidfowl** [comment]
@scalablecory can you elaborate?

**Comment by scalablecory** [comment]
@davidfowl It's the `IAsyncEnumerable` use. Something like this seems a better way to accomplish the same thing:

```c#
var timer = new Timer(delay);

while(await timer.WaitNextPeriodAsync())
{
}
```

If you wouldn't ever `.ToArrayAsync()` on it, then it doesn't fit `IAsyncEnumerable`. I can't imagine ever passing this timer enumerable to anything else (like LINQ) that operates on an `IAsyncEnumerable` -- if I were ever doing that, Rx would probably be a better choice for such an event-based architecture. `foreach`ing a timer also seems weird.

**Comment by scalablecory** [comment]
I'd also want us to look at solving the problems you called out in the existing API as well. It is very common to need an interval timer but not have a neat sequence point to `await` at -- instead, callbacks that happen in parallel to your bigger machine are exactly what you need.

**Comment by davidfowl** [comment]
I disagree with the assertion that foreaching this is weird. The difference between Rx and Ix is push vs pull, nothing else. In fact the pattern you show is just a GetAsyncEnumerator().MoveNextAsync().

Also looking around, golang has a very similar feature called a Ticker that exposes a channel of ticks. It’s a pretty elegant pattern IMHO 😬. I’m it sure why ToArray matters or not. 

> I'd also want us to look at solving the problems you called out in the existing API as well. It is very common to need an interval timer but not have a neat sequence point to await at -- instead, callbacks that happen in parallel to your bigger machine are exactly what you need.

Sure we can do that at the cost of adding more complexity to the existing API (like passing more options to the existing timer ctor?). That’s come up in the past as a potential problem.

**Comment by svick** [comment]
@davidfowl 

> In fact the pattern you show is just a GetAsyncEnumerator().MoveNextAsync().

I think the problem is that when you have an `IEnumerable<T>` or `IAsyncEnumerable<T>`, you're basically saying "you should use `foreach`". If it's expected that you're commonly going to be calling `MoveNextAsync` manually, then it's probably better to be explicit about it by using something like `WaitNextPeriodAsync`.

**Comment by benaadams** [comment]
>  you're basically saying "you should use foreach".

Generalising more its an event pattern; which you couldn't do previously with synchronous enumerators, but now can with asyncenumerators:
```csharp
await foreach (var event in myEventSource)
{

}
```

**Comment by davidfowl** [comment]
> I think the problem is that when you have an IEnumerable<T> or IAsyncEnumerable<T>, you're basically saying "you should use foreach". If it's expected that you're commonly going to be calling MoveNextAsync manually, then it's probably better to be explicit about it by having something like WaitNextPeriodAsync instead.

> Generalising more its an event pattern; which you couldn't do previously with synchronous enumerators, but now can with asyncenumerators:

This is *exactly* the point. You should use foreach. IAsyncEnumerable + await foreach is a new paradigm for C#.

**Comment by svick** [comment]
@davidfowl In my opinion, it's a paradigm that does not fit well here, so it shouldn't be used.

I think with `IAsyncEnumerable<T>`, the important part is the data and the fact that you sometimes have to wait for the next item is mostly incidental. But here, the waiting is what you want and you can invent some data to go along with it, but it's completely incidental.

This can be also seen in the syntax: `await foreach` emphasizes the data and hides the mechanism of retrieving it (`await MoveNextAsync()`), but that's the opposite of what you want here: you don't care about the data, which is visible, but you do care about the `await MoveNextAsync()`, which is hidden.

One more way this surfaces is in how you would document this API. For `WaitNextPeriodAsync()`, it would be roughly:

> Here's this one method, `await` it when you want to wait for the next timer tick. Surround it with whatever syntax you want.

For `IAsyncEnumerable<T>`, it would be (exaggerated for effect):

> It's a sequence of … timestamps, I guess? Though it doesn't behave like most other sequences, since it changes depending on how quickly you iterate it. And maybe iterating it using `await foreach`, like you normally would, doesn't fit your use case, in which case you should `await` `MoveNextAsync` yourself.

To sum up, I think that `IAsyncEnumerable<T>` doesn't fit well logically, its syntax focuses on the wrong thing and it makes it harder to explain what the behavior is. Which is why it shouldn't be used here.

**Comment by benaadams** [comment]
Don't think DateTime/TimeSpan have the precision for this; but another example which `WaitNextPeriodAsync()` by itself doesn't capture (why do you want to tick without knowing the time?)
```csharp
const int framesPerSecond = 60;

var desiredFrameRate = TimeSpan.FromSeconds(1.0 / framesPerSecond);
var ticker = Timer.CreateTimer(desiredFrameRate, desiredFrameRate);

var current = DateTime.UtcNow;
await foreach(var time in ticker)
{
    var delta = time - current;
    current = time;
    
    Console.WriteLine($"Elasped: {delta}, fps: {(double)TimeSpan.TicksPerSecond / delta.Ticks:N2}");
}
```


However, I don't think the data types have the precision for the above sample as 
```csharp
var current = DateTime.UtcNow;
var time = current + TimeSpan.FromSeconds(1.0 / 60);

var delta = time - current;
```
Outputs:
```
Elasped: 00:00:00.0170000, fps: 58.82
```
Rather than
```
Elasped: 00:00:00.0166667, fps: 60.00
```

**Comment by scalablecory** [comment]
@benaadams That use seems problematic, though, doesn't it?

For render lerping it would be more useful to get the time inside the loop (after your async continuation has started running) rather than have `time` be the moment the timer scheduled the continuation. `QueryPerformanceCounter`/etc.

For physics / collision detection it would be more useful to have fixed intervals, not dynamic.

You would also not want it to skip overlapping intervals as the issue desires as that'd halve your framerate, though we could always add a flag to control that feature.

> However, I don't think the data types have the precision for the above

`TimeSpan` has precision down to 1e-7 seconds; I assume you're seeing scheduling delay?

**Comment by scalablecory** [comment]
> I disagree with the assertion that foreaching this is weird.

I think @benaadams' [example](https://github.com/dotnet/corefx/issues/42645#issuecomment-557041421) convinced me that it's not entirely bad, though I still think it's weird.

> The difference between Rx and Ix is push vs pull, nothing else.

I think that's an over-simplification by looking only at their abstraction interfaces, not what they're currently used or the frameworks built around them -- there's a very large difference between the two. Rx is a rich event-driven programming model, whereas enumerables are currently only used for iterating data.

> In fact the pattern you show is just a GetAsyncEnumerator().MoveNextAsync().

Indeed. It's not the basic idea I'm concerned with, but rather specifically the use of `IAsyncEnumerable`.

Right now I feel like enumerables are very data-focused, as @svick said (more eloquently than me). What other APIs in corefx do we have as an example of using enumerables for non-data work? Doing something new isn't a bad thing, I just want to make sure it's clear if we're doing something new as it would set the direction for our user's coding conventions. Enumerables currently have a very predictable usage and I'm really against polluting that by having to think about if an enumerable is there for scheduling or for data

I know that Unity uses `IEnumerable` with `yield return` for ticking entity state, so there is some prior work here at least outside of corefx.

**Comment by omariom** [comment]
What if `WaitNextPeriodAsync` was a public method  and IAsyncEnumerable was exposed via a method (AsEnumerable?) or explicit interface implementation?

**Comment by Clockwork-Muse** [comment]
You **absolutely do not want** something like `DateTime.UtcNow` to drive rapidly firing (or potentially most) timers.  Because even though it's an absolute stamp, it still falls victim to clock slews when trying to set itself via NTP (or from user updates).  It's nonsensical in most of those cases anyways, because you don't actually care about "world time", but some sort of absolute "process time".

@scalablecory is right - you want `QueryPerformanceCounter`, or an equivalent.

Leave `DateTime.UtcNow` for use in full-featured schedulers.

**Comment by benaadams** [comment]
> You absolutely do not want something like DateTime.UtcNow to drive rapidly firing (or potentially most) timers...

It was an example; as you wouldn't use the timer to drive something that needed an 16ms tick as its not fast or precise enough.

I was demonstrating the pattern, which for example listening to mouse moves you could also do as an AsyncEnumerable
```csharp
await foreach(var mouseMove in mouseMoveEvents)
{

}
```

**Comment by benaadams** [comment]
Or perhaps a better example is; but similar vein is button clicks. AsyncEnumerable enable async event handling without `async void` and the horrors that brings.
```csharp
await foreach(var click in btn.ClickEvents)
{
    btn.Enabled = false;
    
    await SomeAsyncMethod();
    
    btn.Enabled = true;
}
```

**Comment by ankitbko** [comment]
Let me see if I can put a different perspective here.

`IEnumerable` (and its async counterpart) always meant I have a collection of *items* that I have freedom to iterate over. The items (value or `Task`) in the collection already exists and I can iterate over them as I want. The `AsyncEnumerable` just means while iterating I may have to asynchronously *wait* (incidentally as @svick put it) for a small amount of time to get the item. But in neither case I expect an *aspect* of scheduling to come while iterating.  This is also represented in fact that `MoveNextAsync` returns `ValueTask` and not `Task`. From [docs](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.valuetask-1?view=netcore-3.0#remarks) - 

> A method may return an instance of this (`ValueTask`) value type when it's likely that the result of its operation will be available synchronously, and when it's expected to be invoked so frequently that the cost of allocating a new Task<TResult> for each call will be prohibitive.

A scheduler (or `Interval`) conceptually is different from iterator (a collection of values). A scheduler *informs/trigger/send* (purposefully avoided *push* to not confuse with rx vs ix) the *subscriber* when the tick occurs. This pattern is essentially same as of Rx/Channel/Observable-Subscriber and the difference between this and enumerable is more than just push vs pull. 

The `Timer` being discussed falls into the scheduler category. This is also represented by the fact that we are proposing to use `Channel` (which is Observer-Observable pattern) to create the timer and then wrapping it up under facade of Enumerator. 

An `IAsyncEnumerable<DateTime>` means I *have* a collection of `DateTime` and it may take me an itsy-bitsy of time to get the items. It does not mean that the `DateTime` values will be available sometime in *future* and I have wait for it to be available. It is for this purpose that we have Observable pattern (Rx).

---

Another way to put is by understanding the thought process of average developer if he gets an an enumerable (exaggerated ofcourse) - 

```cshap

public Task<Something> AwesomeMethod(IAsyncEnumerable<DateTime> foo, other random params)
{
  // Oh look, I get an enumerable of DateTime, this is so great. 
  // I can iterate over it as and when I want. 
  // I am never going to expect this enumerable to behave any 
  // differently than any other enumerable of DateTime (or anything else for that matter).
  // I am definitely never going to expect schedulers and intervals and timers 
  // coming into the mix when I am iterating.
}

```

**Comment by davidfowl** [comment]
Yea I don’t think IAsyncEnumerable is data focused, that’s just something, Linq over events will be a thing (or is a thing depending on who you talk to) and this is just the beginning. 

Modeling multi fire events as IAsyncEnumerable is pretty elegant.

**Comment by davidfowl** [comment]
I’m not sure what schedulers have to do with the conversation but IAsyncEnumerable is the pull and IObservable is the push model. That’s it. Anything else you may have associated with those types aren’t fundamental, they were just assumptions or expectations on how they would behave based on experience with IEnumerable.

**Comment by davidfowl** [comment]
FWIW paradigm shifts always feel uncomfortable,l let’s all suspend disbelief for a second and imagine a world where events were IAsyncEnumerable 😁. What problems would exist, what patterns would emerge as a result? What pitfalls would exist?

That’s the discussion I want to have (along with API approval of course).

**Comment by benaadams** [comment]
> The `AsyncEnumerable` just means while iterating I may have to asynchronously wait (incidentally as @svick put it) for a small amount of time to get the item. But in neither case I expect an aspect of scheduling to come while iterating.

`AsyncEnumerable` is iterating values that may be "not available" yet; rather than `IEnumerable` which is iterating values you already have (though they may be via transform/calculation e.g. Linq). 

Values "not available" yet also encompasses values that do not exist yet as they come from the future; e.g. events. 

Another example would be requests in a http server:
```csharp
await foreach(var request in http.Requests)
{
    await ProcessRequestAsync(request);
}
```

`AsyncEnumerable` is essentially iterating the future. In that vein, a timer based source would be equivalent to `Enumerable.Range` that exists for `IEnumerable`.

>  This is also represented in fact that `MoveNextAsync` returns `ValueTask` and not `Task`.

That also encompasses the fact that Task always allocates per result, and AsyncEnumerable only needs to allocate itself and from that it can generate infinite results without allocating even if it suspends waiting for results.

>  there are examples on other platforms that support this pattern.

Indeed in Dart their async enumerables are called Streams and they support `await for in ` to achieve the same effect.

Stream even has a [`Stream<T>.periodic`](https://api.dart.dev/stable/2.6.1/dart-async/Stream/Stream.periodic.html) constructor to achieve a similar effect as the timer, however it goes further with the generic.

Thinking on the proposed api and Linq integration; it should also take a cancellation, so you could cancel the timer after something like `.Take(50)`.

Also I like the generic approach from Dart; which could be encompassed with `Func<T>` and maybe its name is better? Also maybe a `Func<(T value, bool isComplete)>` so it can indicate to stop producing values and cancel the timer:

```csharp
public class Timer
{
    public static IAsyncEnumerable<T> Periodic(
        Func<T> valueFactory,
        TimeSpan dueTime, 
        TimeSpan period, 
        CancellationToken cancellationToken = default);

    public static IAsyncEnumerable<T> Periodic(
        Func<(T value, bool isComplete)> boundedValueFactory,
        TimeSpan dueTime, 
        TimeSpan period, 
        CancellationToken cancellationToken = default);
}
```

**Comment by benaadams** [comment]
Further to the above 

Then you could build for example a Utc Timer using the `Func<DateTime>`
```csharp
var timer = Timer.Periodic(() => DateTime.UtcNow, TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(1);
await foreach (DateTime time in timer)
{
    Console.WriteLine(time);
}
```
Or the elapsed version with `Func<TimeSpan>`
```csharp
var current = DateTime.MaxValue;
Func<TimeSpan> source = () => {
    var time = DateTime.UtcNow;
    var delta = time - current;
    if (delta < TimeSpan.Zero) delta = TimeSpan.Zero;
    current = time;
    return delta;
};

var timer = Timer.Periodic(source, TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(1);
await foreach (TimeSpan delta in timer)
{
    Console.WriteLine(delta);
}
```

**Comment by ankitbko** [comment]
If the plan is to have events as `IAsyncEnumerable`, would it make sense to have a LINQ method `Interval<T>(this IAsyncEnumerable<T> foo, TimeSpan period)`?
